File: src/csharp/Intel/Iced/Intel/BlockEncoderInternal/TargetInstr.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && BLOCK_ENCODER
namespace Iced.Intel.BlockEncoderInternal {
	readonly struct TargetInstr {
		readonly Instr? instruction;
		readonly ulong address;

		public TargetInstr(Instr instruction) {
			this.instruction = instruction;
			address = 0;
		}

		public TargetInstr(ulong address) {
			instruction = null;
			this.address = address;
		}

		public bool IsInBlock(Block block) => instruction?.Block == block;

		public ulong GetAddress() {
			var instruction = this.instruction;
			if (instruction is null)
				return address;
			return instruction.IP;
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/enc/TargetInstr.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.enc;

final class TargetInstr {
	final Instr instruction;
	final long address;

	public TargetInstr(Instr instruction) {
		this.instruction = instruction;
		address = 0;
	}

	public TargetInstr(long address) {
		instruction = null;
		this.address = address;
	}

	public boolean isInBlock(Block block) {
		return instruction != null && instruction.block == block;
	}

	public long getAddress() {
		Instr instruction = this.instruction;
		if (instruction == null)
			return address;
		return instruction.ip;
	}
}
